src/Functional/functions.php

```diff
@@ -800,3 +803,37 @@ function sequence($monads)
 {
     return traverse(identity, toTraversable($monads));
 }
+
+/**
+ * filterM :: Monad m => (a -> m Bool) -> [a] -> m [a]
+ *
+ * @param callable $f (a -> m Bool)
+ * @param array|Traversable $collection [a]
+ * @return Monad m [a]
+ */
+function filterM(callable $f, $collection)
+{
+    /** @var Monad $monad */
+    $monad = $f(head($collection));
+
+    $_filterM = function($collection) use($monad, $f, &$_filterM){
```

Comment:
There are some techniques but not really readable one, this is usually how you can do recursive function calls using anonymous function in PHP.

The problem is that the variables are bound at the closure creation except when passed by reference and since the function is created before it gets assigned to the variable, there is no other way.

See http://stackoverflow.com/questions/2480179/anonymous-recursive-php-functions